Add random pick code

RoboticsYY opened the pull request:
[PR#1](https://github.com/ros-planning/moveit_example_apps/pull/1) was closed for splitting the huge PR into smaller ones. We would like to split the previous PR#1 into three PRs.

This is the third one. In the previous [PR#2](https://github.com/ros-planning/moveit_example_apps/pull/2) and [PR#3](https://github.com/ros-planning/moveit_example_apps/pull/3), we added the moveit config package and tutorial for the random_pick application. 

This PR intends to add:
- Random pick application code.
- A recorded point cloud file (.pcd) for running the application in simulation.
- Necessary .launch files for bring up the application.
- .travis.yml file and related .sh file to enable CI with moveit_ci and moveit:master-source docker image. 

CI with a custom docker image and moveit_ci is still in development.

RoboticsYY committed:
Add random_pick app control files

RoboticsYY committed:
Add random_pick app gazebo files

RoboticsYY committed:
Add random_pick app visualization and rviz files

RoboticsYY committed:
Add random_pick app source file and point cloud data (.pcd)

RoboticsYY committed:
Add random_pick app demo launch files

RoboticsYY committed:
Update random_pick code package meta info

RoboticsYY committed:
Update .travis.yml

davetcoleman commented:
@mlautman i need some help reviewing this!

mlautman commented on random_pick_depends.sh:
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+export DEPEND=/root/depends
```
I would never want to run this on my machine. It tries to automate too much. It would be better to add these steps to a README and let the user decide how to setup their environment

mlautman replied:
Ok, I see now, this is meant to be run on travis not on a user's machine. In that case, I would recommend cleaning this up a bit so that it is more readable and maintainable and renaming it `travis_before_script.sh` It may also make sense to put it in a subfolder `.travis/before_script.sh` to make it extremely clear to a user that this is not meant to be run on their local machine

RoboticsYY replied:
Truly, I only use this file to travis. Great idea to put it under `.travis` subfolder! Now, the file locates in `.travis/random_pick/`. I think maybe each application can provide a similar file under `.travis/<app_name>/`. In addition, I'm also thinking maybe each application can have its own docker file under `.docker/<app_name>` folder.

mlautman commented on examples/random_pick/random_pick/launch/demo/gpd.launch:
```diff
@@ -0,0 +1,77 @@
+    <arg name="node" value="detect_grasps" />
+  </include>
+  
+	<node name="detect_grasps" pkg="gpd" type="detect_grasps" output="screen">
```
Better to not mix tabs and spaces. Convert to spaces

RoboticsYY replied:
The tabs have been converted to spaces.

mlautman commented on examples/random_pick/random_pick/launch/demo/gpd.launch:
```diff
@@ -0,0 +1,77 @@
+    <param name="samples_topic" value="" />
+    
+    <!-- Plotting parameters -->
+    <param name="plot_normals" value="false" />
```
Nit: These would be good candidates to put in a yaml

RoboticsYY replied:
Great idea! The parameters have been moved to `gpd.yaml` file.

mlautman commented on examples/random_pick/random_pick/launch/demo/hand_geometry.launch:
```diff
@@ -0,0 +1,17 @@
+  <arg name="node" />
+  
+  <param name="$(arg node)/finger_width" value="0.005" />
+<!-- <param name="$(arg node)/hand_outer_diameter" value="0.09" /> -->
```
remove dead code

RoboticsYY replied:
Done.

mlautman reviewed (commented):
Thanks for putting so much work into this! If you could put the setup instructions into a README along with some usage information that would go a long way towards making it easy for the community to learn from this example.

RoboticsYY commented:
@mlautman Thanks for reviewing this PR! I have put the setup instructions to #3. You can check its built effect [here](https://roboticsyy.github.io/moveit_example_apps/doc/random_pick/random_pick.html). It was designed that each application should prepare a tutorial for the community to easily replicate the result. For writing the tutorial, we prepared a [template](https://roboticsyy.github.io/moveit_example_apps/doc/template.html) for each application to follow.

rhaschke committed:
WIP: temporarily use patched moveit_ci

This should be reverted if the corresponding PR https://github.com/ros-planning/moveit_ci/pull/84 is merged.

RoboticsYY committed:
Add Dockerfile

RoboticsYY committed:
Update before_script.sh for random_pick app

RoboticsYY committed:
Fix moveit_ci version

RoboticsYY committed:
Use custom docker image

RoboticsYY commented:
@rhaschke Thanks for fixing the CI issue!

@mlautman I have added a Dockerfile for dealing with the external dependencies of this application. This application has the following components for enabling the CI to align with the moveit_ci framework:
- `.docker/random_pick/Dockerfile`: for setting up the dependencies of external libraries, which overlay on `moveit/moveit:master-source`.
- `.rosinstall/random_pick/random_pick.rosinstall`: for the source dependencies required to be built in the catkin workspace.
- `.travis/random_pick/before_script.sh`: for setting up the environment variables and directory path of dependencies in the docker container.
- a custom docker image, built from the docker file. Perhaps this image can be tagged as `moveit/moveit_example_apps:random_pick`. But now it is only under my docker account.

I'm also thinking maybe other kinds of application can also provide similar components.

mlautman commented on .travis.yml:
```diff
@@ -16,7 +16,11 @@ env:
-    - DOCKER_IMAGE=moveit/moveit:master-source
+      UPSTREAM_WORKSPACE=.rosinstall/random_pick/random_pick.rosinstall
+
+    - DOCKER_IMAGE=xiaotaoqi/moveit_example_apps:random_pick
```
Do we need this separate docker or can this inherit from the moveit ones with some additional Docker commands on top?

RoboticsYY replied:
Sometimes CI times out if directly builds from the moveit ones, so I use this docker image to reduce the CI build time. It is built from the docker file `.docker/random_pick/Dockerfile`, which also builds on top of the moveit-source docker image. I hope new application could also provide their docker file. If the docker files of this repo can build automatically, I'm thinking this docker image could be replaced by the `moveit/moveit_example_apps:random_pick`.

mlautman commented on examples/random_pick/random_pick/src/random_pick.cpp:
```diff
@@ -0,0 +1,501 @@
+
+static const char target_frame_id[] = "base";
+/* place position in the target_frame_id*/
+static std::vector<double> place_position = { -0.45, -0.30, 0.25 };
```
It would be nice if this was a rosparam loaded in with rosparam_shortcuts.

RoboticsYY replied:
Done. These values are now loaded from `random_pick/launch/demo/random_pick.yaml` into rosparam.

mlautman commented on examples/random_pick/random_pick/src/random_pick.cpp:
```diff
@@ -0,0 +1,501 @@
+/* place position in the target_frame_id*/
+static std::vector<double> place_position = { -0.45, -0.30, 0.25 };
+/* place position in joint values*/
+static std::vector<double> joint_values_place = { 0.3857, -1.3737, 1.8270, -2.0249, -1.5708, 1.9569 };
```
same for all these other hardcoded values.

RoboticsYY replied:
Done. Changed them into rosparams.

mlautman commented:
@RoboticsYY I merged the other two PR's if you remove merge conflicts on this one I can merge it as well.

RoboticsYY committed:
Fix catkin_lint error

RoboticsYY committed:
Fix broken links

RoboticsYY committed:
Transform hard-coded values to read from rosparam

RoboticsYY committed:
Update README about how to enable moveit_ci

RoboticsYY commented:
@mlautman Thanks for reviewing these PRs. I have solved the merge conflicts of this PR. 

I use Travis to deploy the built documents to the gh-pages branch in the same way as moveit_tutorials. I may need your help to set up a `$GITHUB_TOKEN` to this repo on Travis.
